### PR TITLE
Avoid issues with signatures that cover any nonce

### LINKED ISSUE #2490: Unprompted Auth: Freshness & Signature Binding
In the current draft, the header is composed as: 
```
{key_id}, {sig_alg}, {hash_alg}, {signature_over_nonce}
```
The server computes the nonce from the TLS key material exporter and verifies the signature. 

This doesn't actually guarantee freshness, because its possible for parties to construct signatures which are valid for any message/nonce. This means that a party which does not have access to the TLS Export material can still inject a valid Auth header. I don't think that's a major problem, but it is a bit unintuitive and its hard to predict how drafts like this will eventually get deployed. 

A simple fix would be in to add an additional parameter, say `v` which is `hash(seperator || nonce)` or similar and have servers check its been correctly computed. This ensures that only parties that can access TLS key material can inject valid headers. 

Behind the scenes, this attack / fix works because hash functions have to be collision resistant for all inputs, whereas signatures don't have to satisfy **any** properties if the associated public key was generated maliciously. [Seems Legit](https://dennis-jackson.uk/assets/pdfs/signatures.pdf) has a longer discussion.

### PR DESCRIPTION
**Pull request by DavidSchinazi**
Fixes #2490

### CODE REVIEW

File: draft-ietf-httpbis-unprompted-auth.md
Comment (martinthomson) [inline draft-ietf-httpbis-unprompted-auth.md:234]: I think that you want to be clearer about what is going on here.  In prose.  That is, 48 bytes is output, then the first 32 bytes is input to the signature and the next 16 bytes are sent along with the signature (to allow the recipient to confirm that the exporter produces the right values).

@dennisjackson, could you just transmit half the nonce instead of extending the KDF output?  The other half should be hard to guess, but confirming that way would save an extra invocation of HMAC.  (I'm not particularly concerned about the cost; I'm just asking to improve my understanding.)

File: draft-ietf-httpbis-unprompted-auth.md
Comment (martinthomson) [inline draft-ietf-httpbis-unprompted-auth.md:244]: I don't like this word, as we've previously discussed.  "Signature Input" would work.

File: draft-ietf-httpbis-unprompted-auth.md
Comment (chris-wood) [inline draft-ietf-httpbis-unprompted-auth.md:279]: Why not simply send the nonce on the wire? The server could then check that (1) the nonce is as expected and (2) the signature is valid. I realize that would result in more bytes on the wire, but splitting up the nonce/verification just seems odd to me. The `verification` value is like a proxy for the actual message being signed, so I thought to myself, why not just send the actual message that's signed?

**Review by chris-wood — CHANGES_REQUESTED** [review]

File: draft-ietf-httpbis-unprompted-auth.md
Comment (chris-wood) [reply draft-ietf-httpbis-unprompted-auth.md:234]: Why not just send the entire nonce?

Comment (DavidSchinazi) [reply draft-ietf-httpbis-unprompted-auth.md:234]: If we send the entire nonce, then a lazy implementer could use that as the value for signature checking instead of properly computing the key exporter. The current design (or MT's proposal to only send part of the nonce) avoid that risk.

My gut feeling is that using another part of the key exporter would make security analysis easier than using part of the nonce, but I'll let @dennisjackson comment on that

File: draft-ietf-httpbis-unprompted-auth.md
Comment (DavidSchinazi) [reply draft-ietf-httpbis-unprompted-auth.md:244]: Signature Input sounds good to me. I'll address this as an editorial change after I land the current train of PRs to avoid having to deal with 17 merge conflicts

File: draft-ietf-httpbis-unprompted-auth.md
Comment (DavidSchinazi) [reply draft-ietf-httpbis-unprompted-auth.md:279]: See other comment thread

File: draft-ietf-httpbis-unprompted-auth.md
Comment (chris-wood) [reply draft-ietf-httpbis-unprompted-auth.md:234]: I'm not sure I buy the lazy implementer argument, but I also think sending half the nonce would be sufficient here.

Comment (DavidSchinazi) [reply draft-ietf-httpbis-unprompted-auth.md:234]: I agree that sending half the nonce would work, but what's the benefit? We need to transmit some sort of verification bit regardless, so the only benefit I see to reusing the start of the nonce instead of getting some more key exporter output is that it saves an HKDP operation. I'd rather take the minimal CPU hit to avoid bit reuse if there's no other benefit

Comment (chris-wood) [reply draft-ietf-httpbis-unprompted-auth.md:234]: It just seems like the most obvious thing to do: send the message that is being signed.

File: draft-ietf-httpbis-unprompted-auth.md
Comment (DavidSchinazi) [reply draft-ietf-httpbis-unprompted-auth.md:244]: Filed #2578 to track this

File: draft-ietf-httpbis-unprompted-auth.md
Comment (dennisjackson) [reply draft-ietf-httpbis-unprompted-auth.md]: Sorry for the slow response. 

Short answer: @chris-wood is right that from a formal analysis perspective sending the entire message or a hash of the entire message would also be fine and this is the generic fix to this kind of signature attack. @martinthomson's suggestion of sending half the signature input also works in this specific case. However, for pragmatic reasons I think @DavidSchinazi's solution is nicest overall. 

Longer answer: 

From the perspective of a formal analysis, using the full or half signature input is fine. The positive property we're looking for is that the sender is committed to a single message input, so the full message obviously achieves that. The half input is only going to match with prob 2^128 so also fine. Thinking about negative consequences, we're going from transmitting the signature of the authenticator to the authenticator itself. In a formal analysis, this isn't ever going to be a problem because signature schemes are assumed to be unforgeable (EUF-CMA) but nothing more, so an attacker in 'proof world' can do things like take a signature and reverse it to learn the message (even though that's typically hard in the real world).   

Pragmatically though, I like @DavidSchinazi's solution best. It lessens the risk of a lazy implementer using the sender's value by mistake in the signature verification or trying to save a `memcmp` by skipping the second check of the signature input. 

Another *very* minor benefit that in some signature schemes like Ed25519, the attacker can't reverse it in practice as they only learn `h(pk || m)` and in this case `m` is high entropy. So in the case the implementer uses Ed25519, by using the extra KDF output you make it intractable for an attacker who has witnessed an unprompted auth header to extract the authenticator value and make their own valid header with their key.

**Pull request merged**
